keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,730 @@
+3. For the GA release, the CRI implementation should be the source of truth for all pod and container level metrics that external parties rely on (no matter how many endpoints the Kubelet advertises).
+
+
+### Risks and Mitigations
+
+- To properly move to CRI stats, it is likely there are some metrics that we'll want to not support (Accelerator/UserDefined). We should be careful to not break entities that rely on these metrics.
+- A large part of this work is changing the source of the Summary API metrics. In doing so, there is a risk that collecting from a new source will change how the stats look in aggregate (and risk bugs popping up in new areas).
+- cAdvisor has a long history of collecting these stats. There is a risk that changing the source of the stats to the CRI implementation can cause performance regressions
+as cAdvisor is fine tuned to perform in an adequate manner.
+    - CRI implementations should do performance regression analyses to ensure the change does not regress too much.
+
+## Design Details
+
+
+#### CRI Implementation
+The CRI implementation will need to be extended to support reporting the full set of container-level from the [Summary API][#summary-container-stats-object].
+
+##### ContainerStats additions
+Currently, the CRI endpoints `{,List}ContainerStats` report the following fields for each container:
+- CPU
+    - `usage_core_nano_seconds`
+- Memory
+    - `working_set_bytes`
+- Filesystem
+    - `inodes_used`
+    - `used_bytes`
+
+These correspond to some fields of the [ContainerStats][#summary-container-stats-object] object reported by cAdvisor. Each of the CRI Stats fields will have to be extended to support the remaining fields. Here are some concrete additions the CRI will need, as well as notes showing what fields correspond to what between CRI and cAdvisor (note: a table version of this can be seen above):
+```
+-// CpuUsage provides the CPU usage information.
+-message CpuUsage {
+-    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
++    // Corresponds to Stats Summary API CPUStats Time field
+-    int64 timestamp = 1;
+-    // Cumulative CPU usage (sum across all cores) since object creation.
++    // Corresponds to Stats Summary API CPUStats UsageCoreNanoSeconds
+-    UInt64Value usage_core_nano_seconds = 2;
++    // Total CPU usage (sum of all cores) averaged over the sample window.
++    UInt64Value usage_nano_seconds = 3;
+-}
+
+-// MemoryUsage provides the memory usage information.
+-message MemoryUsage {
+-    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
++    // Corresponds to Stats Summary API MemoryStats Time field
+-    int64 timestamp = 1;
+-    // The amount of working set memory in bytes.
++    // Corresponds to Stats Summary API MemoryStats WorkingSetBytes field
+-    UInt64Value working_set_bytes = 2;
```

Comment:
i am actually curious how this field is handled on windows.  it is a confusing field to calculate on linux.